### PR TITLE
Resolve #245: Multiple template engines for components.

### LINKED ISSUE #245: Multiple template engines support in a project
There was as feature request by @ptz for supporting several template engines inside one project. It would be convenient if you have several teams who create components in the shared repo and use different template engines.

# Proposal

* Every template provider must implment a `getName()` method that returns a string which is used as a unique identifier of the template engine. In case of name collision the last registered wins.
* Every template provider must implement a `isTemplateSupported(filename, templateString)` which is an indicator if the template provider is able to compile the template
* Catberry compiles a template using the first template provider that responds with `true`
* Catberry stores the name returned by `getName` in a component specification (appDefinitions) this is how it knows which templateProvider to use in a browser to render a template.

Following official template providers have to be updated in order to support new functionality (they will keep backward compatibility though):
- [ ] catberry
- [x] [catberry-handlebars](https://github.com/catberry/catberry-handlebars)
- [x] [catberry-pug](https://github.com/catberry/catberry-pug)
- [x] [catberry-dust](https://github.com/catberry/catberry-dust)

### PR DESCRIPTION
**Pull request by pragmader**
Now it's possible to register several template engines and use different
template across component and even inside a component for main and error templates.

### CODE REVIEW

File: browser/loaders/ComponentLoader.js [hunk 1]
```diff
@@ -33,11 +34,16 @@ class ComponentLoader extends LoaderBase {
...
+		 * @type {Object}
 		 * @private
 		 */
-		this._templateProvider = locator.resolve('templateProvider');
+		this._templateProvidersByNames = templateHelper
+			.resolveTemplateProviders(locator)
+			.reduce((map, current) => {
+				map[current.getName()] = current;
+				return map;
+			}, Object.create(null));
```
Comment (kalyanov) [inline browser/loaders/ComponentLoader.js]: Perhaps we have to realise it in `templateHelper` as `resolveTemplateProvidersByNames()`

File: lib/helpers/loadHelper.js [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+		const errorTemplateSourcePromise = Promise.resolve()
+			.then(() => {
+				component.errorTemplateSource = null;
+				const relativePath = component.properties.errorTemplate;
+				if (typeof (relativePath) !== 'string') {
+					return null;
+				}
+				const templatePath = path.resolve(
+					path.dirname(component.path),
+					component.properties.errorTemplate
```
Comment (kalyanov) [inline lib/helpers/loadHelper.js]: I think, we can use `relativePath` path const here

File: lib/helpers/loadHelper.js [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+/**
+ * Finds a template provider that would work with the template.
+ * @param {Array<TemplateProviders>} providers List of template providers.
+ * @param {string} filename Filename of the template.
+ * @param {string} templateContent Content of the template file.
+ * @returns {TemplateProvider|null} Template provider that supports the template.
+ * @private
+ */
+function findTemplateProvider(providers, filename, templateContent) {
+	var templateProvider = null;
```
Comment (kalyanov) [inline lib/helpers/loadHelper.js]: Maybe `let`

File: browser/loaders/ComponentLoader.js (hunk 1 above)
Comment (pragmader) [inline browser/loaders/ComponentLoader.js]: Good point, agree.

File: lib/helpers/loadHelper.js (hunk 2 above)
Comment (pragmader) [inline lib/helpers/loadHelper.js]: Didn't get it, what do you mean?

File: lib/helpers/loadHelper.js (hunk 3 above)
Comment (pragmader) [inline lib/helpers/loadHelper.js]: I'm not using `let` unless I have to. It causes more generated code going through Babel.

File: lib/helpers/loadHelper.js (hunk 2 above)
Comment (kalyanov) [inline lib/helpers/loadHelper.js]: line 52
`const relativePath = component.properties.errorTemplate;`

File: lib/helpers/loadHelper.js (hunk 3 above)
Comment (kalyanov) [inline lib/helpers/loadHelper.js]: ok

File: lib/helpers/loadHelper.js (hunk 2 above)
Comment (pragmader) [inline lib/helpers/loadHelper.js]: Right, now I see.

**Pull request merged**